compiler/optimizer/Operation.cpp

```diff
@@ -0,0 +1,347 @@
+
+      // long operations
+      case TR::lconst:
+      {
+         result.type = Int64;
+         result.data.lconst = node->getLongInt();
+         break;
+      }
+      case TR::lload:
+      {
+         result = symbolTable[node->getSymbol()];
+         break;
+      }
+      case TR::lloadi:
+      {
+         result.type = TR::TreeInterpreter::DATATYPE::Address;
+         result.data.aconst = (uintptrj_t)&(symbolTable[node->getSymbol()]);
+         break;
+      }
+      case TR::lstore:
+      {
+         symbolTable[node->getSymbol()] = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         break;
+      }
+      case TR::ladd:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         operand2 = nodeValueMap[node->getChild(1)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = operand1.data.lconst + operand2.data.lconst;
+         break;
+      }
+      case TR::lsub:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         operand2 = nodeValueMap[node->getChild(1)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = operand1.data.lconst - operand2.data.lconst;
+         break;
+      }
+      case TR::lmul:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         operand2 = nodeValueMap[node->getChild(1)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = operand1.data.lconst * operand2.data.lconst;
+         break;
+      }
+      case TR::ldiv:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         operand2 = nodeValueMap[node->getChild(1)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = operand1.data.lconst / operand2.data.lconst;
+         break;
+      }
+      case TR::lrem:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         operand2 = nodeValueMap[node->getChild(1)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = operand1.data.lconst % operand2.data.lconst;
+         break;
+      }
+      case TR::lneg:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = ~operand1.data.lconst;
+         break;
+      }
+      case TR::labs:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         result.type = Int64;
+         if (operand1.data.lconst >= 0)
+            result.data.lconst = operand1.data.lconst;
+         else
+            result.data.lconst = -(operand1.data.lconst);
+         break;
+      }
+      case TR::lshl:
+      {
+         operand1 = nodeValueMap[node->getChild(0)->getGlobalIndex()];
+         operand2 = nodeValueMap[node->getChild(1)->getGlobalIndex()];
+         result.type = Int64;
+         result.data.lconst = operand1.data.lconst << operand2.data.lconst;
```

Comment:
For shifts 64 or greater, we now return 0.
How would you suggest dealing with negative shifts?

Reply:
Unfortunately, that's not enough. Because you're using signed integers, you can still run into undefined behaviour if the shift causes signed overflow.

For negative shifts, I would suggest either asserting or throwing an exception.

Reply:
Also, I'm not sure if it makes sense to return 0 for out of range right shifts.

Reply:
Okay, we tried to shield against overflow in a new commit, with asserts.